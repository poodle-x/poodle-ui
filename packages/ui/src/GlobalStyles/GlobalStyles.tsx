import { css, Global } from "@emotion/react";
import React from "react";
import { CSSObjectSystem } from "../styled/";

import {
	getFontBase,
	getThemeValue,
	system,
	ThemeConfig,
	useTheme,
} from "../theme";

export interface GlobalStylesProps {
	theme?: ThemeConfig;
	styles?: CSSObjectSystem;
}

function GlobalStyles(props: GlobalStylesProps) {
	const { theme } = useTheme();

	const { styles } = props;

	const globalStyles = React.useMemo(() => {
		return getThemeValue(
			{
				theme,
			},
			theme?.global?.styles,
			{}
		);
	}, [theme]);

	const base = React.useMemo(() => {
		return system(
			{
				body: {
					padding: 0,
					margin: 0,
					bg: "bg",
					...getFontBase()({
						theme,
					}),
				},
			},
			{
				sendProps: {
					styles,
					theme,
				},
			}
		);
	}, [styles, theme]);

	const extend = React.useMemo(() => {
		return system(
			{
				...globalStyles,
				...styles,
			},
			{
				sendProps: {
					styles,
					theme,
				},
			}
		);
	}, [globalStyles, styles, theme]);

	return <Global styles={css(base, extend as any)} />;
}

export default GlobalStyles;
