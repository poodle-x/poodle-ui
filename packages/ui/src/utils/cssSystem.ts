import { CSSObjectSystem } from "../styled";
import { getThemeValue, system, ThemeValue } from "../theme";

export function cssSystem(data: {
	props: any;
	base: CSSObjectSystem[];
	applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
	override?: ThemeValue<CSSObjectSystem>;
}) {
	const { props, applies = [], override, base } = data;

	const overrideTheme = getThemeValue<CSSObjectSystem | undefined>(
		props,
		override,
		undefined
	);

	if (overrideTheme) {
		return system(
			{ ...overrideTheme, theme: props.theme },
			{
				sendProps: props,
			}
		);
	}

	return [
		...base.map((b) =>
			system(
				{ ...b, theme: props.theme },
				{
					sendProps: props,
				}
			)
		),
		...applies.reduce<(CSSObjectSystem | CSSObjectSystem[])[]>((prev, s) => {
			const v = getThemeValue(props, s, undefined);

			if (v) {
				return [
					...prev,
					system({ ...v, theme: props.theme }, { sendProps: props }),
				];
			}

			return prev;
		}, []),
	];
}
