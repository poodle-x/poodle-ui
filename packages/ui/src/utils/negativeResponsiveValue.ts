import { ResponsiveProp, ThemeValue } from "../theme";

function negativePrimitiveValue(
	value: ThemeValue<string | number>
): ThemeValue<string | number> {
	if (typeof value === "string") {
		return `-${value}`;
	}

	if (typeof value === "number") {
		return value * -1;
	}

	return value;
}

export function negativeResponsiveValue(
	value: ResponsiveProp<number | string>
) {
	if (typeof value === "string") {
		return `-${value}`;
	}

	if (typeof value === "number") {
		return value * -1;
	}

	if (Array.isArray(value)) {
		return value.map((v) => {
			return negativePrimitiveValue(v);
		});
	}

	if (typeof value === "object") {
		return Object.keys(value).reduce<{
			[key: string]:
				| ThemeValue<string | number>
				| ThemeValue<string | number>[];
		}>((r, k) => {
			const v = value[k];
			const s = Array.isArray(v)
				? v.map((vv) => {
						return negativePrimitiveValue(vv);
				  })
				: negativePrimitiveValue(v);
			return {
				...r,
				[k]: s,
			};
		}, {});
	}

	return value;
}
