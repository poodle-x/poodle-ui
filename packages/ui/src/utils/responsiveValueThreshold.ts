import { ResponsiveProp } from "../theme";

export function responsiveValueThreshold<T>(
	value: ResponsiveProp<T>,
	threshold: string,
	breakPointScale: string[]
): ResponsiveProp<T> {
	let calcValue: { [k: string]: ResponsiveProp<T> } | undefined = undefined;

	if (threshold && !Array.isArray(value) && typeof value === "object") {
		const scaleIndex = breakPointScale.findIndex((b) => b === threshold);
		if (scaleIndex >= 0) {
			breakPointScale.slice(scaleIndex, breakPointScale.length).forEach((b) => {
				if (!calcValue) {
					calcValue = {};
				}
				calcValue[b] = (value as any)[b];
			});
		}
	} else if (threshold && Array.isArray(value)) {
		const scaleIndex = breakPointScale.findIndex((b) => b === threshold);
		const br = breakPointScale.slice(scaleIndex, breakPointScale.length);
		if (scaleIndex >= 0) {
			br.forEach((b, i) => {
				if (value[scaleIndex + i + 1]) {
					if (!calcValue) {
						calcValue = {};
					}
					calcValue[b] = value[scaleIndex + i + 1];
				}
			});
		}
	} else if (
		threshold &&
		(typeof value === "string" || typeof value === "number")
	) {
		return {
			[threshold]: value,
		};
	}

	return calcValue || value;
}
