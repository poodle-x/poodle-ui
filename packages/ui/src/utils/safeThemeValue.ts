export function safeThemeUnitValue(
	value: string | number,
	config?: {
		unit?: string;
	}
): string {
	const finalConfig = {
		unit: "px",
		...config,
	};

	if (typeof value === "string") {
		return value;
	}

	return `${value}${finalConfig.unit}`;
}

export default safeThemeUnitValue;
