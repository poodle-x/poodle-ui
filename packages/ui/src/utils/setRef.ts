import React from "react";

export default function setRef(
	ref:
		| ((instance: HTMLElement | null) => void)
		| React.MutableRefObject<HTMLElement | null>
		| null,
	value: HTMLElement | null
) {
	if (typeof ref === "function") {
		ref(value);
	} else if (ref) {
		ref.current = value;
	}
}
