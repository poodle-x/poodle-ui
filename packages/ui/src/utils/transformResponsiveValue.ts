import { ResponsiveProp, ThemeValue } from "../theme";

export function transformResponsiveValue<T, R = T>(
	value: ResponsiveProp<T>,
	transform: (value: ThemeValue<T>) => R
): ResponsiveProp<R> {
	let result: ResponsiveProp<R> = {};
	if (Array.isArray(value)) {
		result = value.map((v) => {
			return transform(v);
		});
	} else if (typeof value === "object") {
		result = Object.keys(value).reduce<{
			[key: string]: ThemeValue<R> | Array<ThemeValue<R>>;
		}>((a, key) => {
			const v: ThemeValue<T> = (value as any)[key];
			if (typeof v === "function") {
				return {
					...a,
					[key]: transform(v),
				};
			} else if (Array.isArray(v)) {
				return {
					...a,
					[key]: v.map((vv) => {
						return transform(vv);
					}),
				};
			}

			return {
				...a,
				[key]: transform(v),
			};
		}, {});
	} else {
		result = transform(value);
	}

	return result;
}
