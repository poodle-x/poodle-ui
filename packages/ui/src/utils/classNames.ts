import React from "react";
import { css, CSSObjectSystem, cx } from "../styled";
import { ThemeValue } from "../theme";
import { cssSystem } from "./cssSystem";

type ClassName =
	| undefined
	| string
	| ((props: any) => string)
	| { [key: string]: boolean };

export function genClassNames(data: {
	props: any;
	classNames: Array<ClassName>;
}) {
	const { props, classNames } = data;

	let resultClassNames = "";

	classNames.forEach((cn) => {
		if (typeof cn === "function") {
			resultClassNames = cx(resultClassNames, cn(props));
			return;
		}

		resultClassNames = cx(resultClassNames, cn);
	});

	return resultClassNames;
}

export function classNameCompose(data: {
	props: any;
	overrideClassName?: string | boolean | undefined;
	styles: {
		base: CSSObjectSystem[];
		applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
		override?: ThemeValue<CSSObjectSystem>;
	};
}) {
	const { overrideClassName, styles, props } = data;

	if (typeof overrideClassName === "string") {
		return overrideClassName;
	}

	if (overrideClassName === true) {
		return "";
	}

	return css(
		cssSystem({
			...styles,
			props,
		})
	);
}

export function genListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
): { [key in keyof typeof lists]: string } {
	const result: { [key: string]: string } = {};
	Object.keys(lists).forEach((key) => {
		let resultClassNames = "";
		const listKeyData = lists[key];
		resultClassNames = genClassNames({
			props: { ...baseProps, ...listKeyData.props },
			classNames: listKeyData.classNames,
		});
		result[key] = resultClassNames;
	});

	return result;
}

export function useGenListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
) {
	return React.useMemo(() => {
		return genListClassNames(baseProps, lists);
	}, [lists, baseProps]);
}
