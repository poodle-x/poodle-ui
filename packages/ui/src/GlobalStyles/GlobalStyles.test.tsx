import React from "react";
import { render } from "../utils/test";
import GlobalStyles from "./GlobalStyles";

describe("<GlobalStyles />", () => {
	it("should render GlobalStyles correctly", () => {
		const { container } = render(<GlobalStyles />);
		expect(container.firstChild).toMatchSnapshot();
	});
});
