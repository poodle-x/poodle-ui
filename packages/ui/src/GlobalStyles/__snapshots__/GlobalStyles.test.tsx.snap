// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalStyles /> should render GlobalStyles correctly 1`] = `null`;
